### PR TITLE
Add Unit Testing Guidelines

### PR DESCRIPTION
**Pull request by omusavi**
This change adds the documentation and guidelines on writing unit tests. This includes subpages on authoring unit tests, mocking, and test driven development

### CODE REVIEW

**Comment by TessFerrandez** [comment]
@omusavi is there an open issue associated with this, can you please link it if so, so it gets closed as part of the PR?

**Comment by omusavi** [comment]
No open issue, this is just the initial work we are doing to update the automated testing sections. Can create one if that is something that is required!

File: automated-testing/unit-testing/tdd_example.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
+# Test-Driven Development Example
+
+With this method, rather than writing all your tests up front, you write one test a time and then switch to write the
+system code that would make that test pass. It's important to write the bare minimum of code necessary even if it is not
+actually "correct". Once the test passes you can refactor the code to make it maybe make more sense, but again the logic
```
Comment (ohthehugemanatee) [inline automated-testing/unit-testing/tdd_example.md:5]: I think the "bare minimum" concept is best illustrated as trying to "cheat" your own test. 

Can you pass it with a simple "print" statement? Then write that print statement. Once you have something stupid that passes the test, improve the test so your cheat doesn't work anymore. Then go back and write the minimal code to pass it agani.

With this back and forth process, you only expand your code by small functional units, and your tests will rigorously push all the capabilities and limits of the target system.

Comment (omusavi) [reply automated-testing/unit-testing/tdd_example.md:5]: that is a great way of putting it! I will update to include that analogy

File: automated-testing/unit-testing/why-unit-tests.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+It is no secret that writing unit tests is hard, and even harder to write well. Writing unit tests also increases the
+development time for every feature. So why should we bother writing them?
+
+## Reduce costs
+
+There is no question that the later a bug is found, the more expensive it is to fix; especially so if the bug makes it
+into production. A [2008 research study by IBM](ftp://ftp.software.ibm.com/software/rational/info/do-more/RAW14109USEN.pdf)
+estimates that a bug caught in production could cost 6 times as much as if it was caught during implementation.
+
+## Increase Developer Confidence
```
Comment (ohthehugemanatee) [inline automated-testing/unit-testing/why-unit-tests.md:12]: I think this is more strongly stated as "Prove your code works (no matter what you change)". 

"Developer confidence" sounds like it's only for the developer. Really, test coverage is for developers, POs, architects, and anyone else with an investment in working code. For a DevOps engineer, it's the security that redeployment doesn't break anything. For POs, it's the confidence that their product (still) has all the capabilities they've paid for. For developers, it's the freedom to do any level of cleanup, refactoring, or rewrite without breaking anything. 

Also, "proving" is a key verb, here. Code with tests is provably working. Code without tests is a mystery.

File: automated-testing/unit-testing/tdd_example.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+should be simple. As you write more tests, the logic gets more and more complex, but you can continue to make the
+minimal changes to the system code with confidence because all code that was written is covered.
```
Comment (ohthehugemanatee) [inline automated-testing/unit-testing/tdd_example.md:7]: Probably worth adding something about the "gradually increasing complexity" rule. Always write the simplest test first. Avoid writing a test for the complex core of your algorithm, and always take the next simplest secondary aspect to test next.

**Comment by ohthehugemanatee** [comment]
Added some comments.

This is great work, and really valuable writing. Thank you for puting in this effort! 

I don't think this level of depth belongs in the playbook, however. I would rather see these published separately externally, and we can link to them as good guidance. 

Remember that the playbook should be easy to digest, and should apply regardless of language or technical area. Testing has a different practical meaning in infrastructure vs data science vs IoT. The rules and strategies presented here are derived from object oriented languages. They don't apply to functional, procedural, or declarative languages. They are incomplete in serverless architectures. etc etc. It's just not practical to cover everything.

I encourage you to publish this excellent work on a blog, medium, or even a standalone github post of your own. In this guide I suggest we stick to the core values, the real fundamentals of testing whch apply to all corners of our discipline. E.g.:

> Tests should: 
>
> * demonstrate that your project works as a whole (end-to-end tests);
> * demonstrate that each individual part of your project works (unit or component tests);
> * demonstrate that each two interacting pieces of your project work together (integration tests);
> * not depend on any external services (if possible);
> * be run as frequently as practical;
> * be a litmus test before merging any code changes.

We can link to good testing examples in different domains. It is not the job of the playbook to teach specific techniques like dependency injection. We have the Internet for that.

**Comment by omusavi** [comment]
Thanks for the review Campbell and I appreciate the discussion. I started off thinking it would be as you described with general guidelines but I found there was a lot to say and I did not want to just say things in the main page without describing them or just linking off to sites that I felt like didn't explain it as well. Moving some of those subpages to a separate page is definitely something we can do (I am curious how the rest of the testing EF team feels if that would be a better way of organizing).

I totally agree that this is focused on strategies of OOP, especially in the authoring section. I think the "How" of writing unit tests is important and specifying different strategies fits within the scope of what we are trying to achieve in the playbook but expanding on doing unit testing in different paradigms is something I would love to expand on. For now I can specify that this is OOP focused and create an issue to expand on that in the future. 

And just a note that all of these are in the unit testing section so while some of it may apply to other testing areas, I am focused only on unit testing.

File: automated-testing/unit-testing/readme.md [hunk 4]
```diff
@@ -1,67 +1,210 @@
...
-Unit testing works in conjunction with integration tests and end-to-end tests for larger pieces of functionality.
+Most languages also have a wide suite of **unit testing frameworks** and test runners. These test frameworks have
+a wide range of functionality, but the base functionality should be a way to organize your tests and run them quickly.
 
-In order to keep execution of unit tests fast and executable as part of a CI/CD pipeline, tests can provide mock implementations of other parts of the application or 3rd party services.
+Finally there is your **unit test code**; unit test code is generally short and simple, preferring repetition to adding
+layers and complexity to the code.
 
-For integration or end-to-end testing, mocks should be replaced with API calls to the system they are simulating.
+## Applying the Unit Testing [the how]
```
Comment (RichCzyzewski) [inline automated-testing/unit-testing/readme.md:39]: Nit: Should have consistent casing between "[The What]" and "[the how]"

File: automated-testing/unit-testing/readme.md [hunk 5]
```diff
@@ -1,67 +1,210 @@
...
+
+#### Test-Driven Development
+
+Test-Driven Development (TDD) is less a technique in how your code is designed, but a technique for writing your
+code that will lead you to a testable design from the start. The basic premise of test-driven development is that you
+write your test code first and then write the system under test to match the test you just wrote. This way all the test
+design is done up front and by the time you finish writing your system code, you are already at 100% test pass rate and
+test coverage. It also guarantees testable design is built into the system since the test was written first!
+
+For more information on TDD and an example, see the page on [Test-Driven Development](./tdd.md)
```
Comment (RichCzyzewski) [inline automated-testing/unit-testing/readme.md:97]: Link doesn't work.

File: automated-testing/unit-testing/readme.md [hunk 6]
```diff
@@ -1,67 +1,210 @@
...
-Good unit tests follow a few general principles:
+### Techniques
 
-- Pass / Fail tests ensure intended succeeds and fails as designed.
-- Transactional tests ensure transactions commit or roll back as designed
-- CRUD operations work as designed
-- All data created for or during a test is localized to the test to allow for parallel test execution
-- All data created during or for a test gets cleaned up after tests completed
+These are some commonly used techniques that will help when authoring unit tests. For some examples, see the pages on
+using [abstraction and dependency injection to author a unit test](authoring_example.md), or how to do [test-driven development](tdd.md)
```
Comment (RichCzyzewski) [inline automated-testing/unit-testing/readme.md:48]: TDD link doesn't work.

File: automated-testing/unit-testing/tdd_example.md [hunk 7]
```diff
@@ -0,0 +1,218 @@
+# Test-Driven Development Example
+
+With this method, rather than writing all your tests up front, you write one test a time and then switch to write the
```
Comment (RichCzyzewski) [inline automated-testing/unit-testing/tdd_example.md:3]: Nit: one test **at** a time

**Review by RichCzyzewski — COMMENTED** [review]
Found a few minor things to fix, but otherwise looks great. Thanks!

**Review by RichCzyzewski — APPROVED** [review]
Approving

File: automated-testing/unit-testing/mocking.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+One of the key components of writing unit tests is to remove the dependencies your system has and replacing it with an
+implementation you control. The most common method people as the replacement for the dependency is a mock, and mocking
+frameworks exist to help make this process easier.
+
+Many frameworks and articles use different meanings for the differences between test doubles. For this sake of clarity,
+this article will be using the [definitions provided by Martin Fowler](https://martinfowler.com/articles/mocksArentStubs.html#TheDifferenceBetweenMocksAndStubs). The most commonly
+used form of test double is Mocks, but there are many cases where Mocks perhaps are not the best choice and Fakes should
+be considered instead.
+
+## Stubs
```
Comment (gled4er) [inline automated-testing/unit-testing/mocking.md:12]: Do you think table comparing all the options will help here?

**Review by gled4er — APPROVED** [review]

**Pull request merged**